server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java

```diff
@@ -99,6 +102,37 @@ public ShuffleServerGrpcService(ShuffleServer shuffleServer) {
     this.shuffleServer = shuffleServer;
   }
 
+  @Override
+  public void unregisterApp(
+      RssProtos.AppUnregisterRequest request,
+      StreamObserver<RssProtos.AppUnregisterResponse> responseStreamObserver) {
+    String appId = request.getAppId();
+
+    StatusCode result = StatusCode.SUCCESS;
+    String responseMessage = "OK";
+    try {
+      Map<Integer, RangeMap<Integer, ShuffleBuffer>> shuffleIdToBuffers =
+          shuffleServer.getShuffleBufferManager().getBufferPool().get(appId);
+      if (shuffleIdToBuffers != null) {
+        HashSet<Integer> shuffleIds = Sets.newHashSet(shuffleIdToBuffers.keySet());
+        shuffleIds.forEach(
+            shuffleId ->
+                shuffleServer.getShuffleTaskManager().removeShuffleDataAsync(appId, shuffleId));
```

Comment:
I think the deleting all shuffleIds  one by one is not a good way.

You could follow the next step to extend
1. create a new method in ShuffleTaskManager, like this:
```
public void removeShuffleDataAsync(String appId) {
    expiredAppIdQueue.add(
        new AppPurgeEvent(appId, getUserByAppId(appId));
  }
```
2. use above method in grpc service.

Reply:
I tried to do this at first, but I found that when processing AppPurgeEvent, it will first determine whether the App has expired. 
```java
 // the thread for clear expired resources

// 1
    Runnable clearResourceRunnable =
        () -> {
          while (true) {
            PurgeEvent event = null;
            try {
              event = expiredAppIdQueue.take();
              long startTime = System.currentTimeMillis();
              if (event instanceof AppPurgeEvent) {
                removeResources(event.getAppId(), true);
                double usedTime =
                    (System.currentTimeMillis() - startTime) / Constants.MILLION_SECONDS_PER_SECOND;
                ShuffleServerMetrics.summaryTotalRemoveResourceTime.observe(usedTime);
              }
// 2
public void removeResources(String appId, boolean checkAppExpired) {
    Lock lock = getAppLock(appId);
    try {
      lock.lock();
      LOG.info("Start remove resource for appId[" + appId + "]");
      if (checkAppExpired && !isAppExpired(appId)) {
        LOG.info(
            "It seems that this appId[{}] has registered a new shuffle, just ignore this AppPurgeEvent event.",
            appId);
        return;
      }

// 3
 private boolean isAppExpired(String appId) {
    if (shuffleTaskInfos.get(appId) == null) {
      return true;
    }
    return System.currentTimeMillis() - shuffleTaskInfos.get(appId).getCurrentTimes()
        > appExpiredWithoutHB;
  }
```

According to my understanding, in this case, the App is not necessarily expired, so it may not be able to play the role of unregister shuffle. If my understanding is wrong, please point me out. 

```java
public void removeShuffleDataAsync(String appId) {
    shuffleTaskInfos.remove(appId);
    expiredAppIdQueue.add(
            new AppPurgeEvent(appId, getUserByAppId(appId));
  }

```
Maybe we can do this(remove the appId first and then do what you said) , WDYT?

Reply:
We should preserve these logic that solved the bug of #1212 . 

How about introducing new event named AppUnregisterPurgeEvent?

Reply:
#1212 This is mainly due to the asynchronous event that once thought the App was no longer used, because re-register will be used again (resend heartbeat), so when removeResource will check again whether the App has expired.

Considering this issue, here, we want to unregister all shuffles of the app at the **"current"** moment, which means that if` AppUnregisterPurgeEvent` is introduced, we need to pass all shuffleIds of the current moment into the event. If so, why don't we just use `ShufflePurgeEvent` and pass the shuffleIds into it.

Reply:
Maybe like this.
```java
public void removeShuffleDataAsync(String appId) {
    Map<Integer, RangeMap<Integer, ShuffleBuffer>> shuffleIdToBuffers =
            this.shuffleBufferManager.getBufferPool().get(appId);
    if (shuffleIdToBuffers != null) {
      HashSet<Integer> shuffleIds = Sets.newHashSet(shuffleIdToBuffers.keySet());
      expiredAppIdQueue.add(new ShufflePurgeEvent(appId, getUserByAppId(appId) , (List<Integer>) shuffleIds);
    }
  }
```

Reply:
> Considering this issue, here, we want to unregister all shuffles of the app at the "current" moment, which means that if AppUnregisterPurgeEvent is introduced, we need to pass all shuffleIds of the current moment into the event. If so, why don't we just use ShufflePurgeEvent and pass the shuffleIds into it.

I means we could remove app without rechecking the heartbeat timeout.

![image](https://github.com/apache/incubator-uniffle/assets/8609142/6e787e5e-a4b6-41dd-9b1b-e300d47e472f)

Reply:
I understand, I have completed the modification, please check it out, thanks!